### DISCUSSION TITLE
Icon coding

### DISCUSSION THREAD

**Discussion by rebel7580** [HASwitchPlate/HASPone#101] [Q&A]
I'm trying to set icons via my own MQTT application. I seem to have problems doing any form of copy/paste, I think because the various editors/debuggers I'm using are not very friendly to the encoding used for the icons. I do have my editor set for UTF-8 just in case.  However, when I try to paste in an icon from the foreAwesome web page, some work, but most don't (nothing shows on a button when sent to the display.  I use an "xstr" command to send to the display. I've copied this xstr command from one sent successfully by Home Assistant, so I'm pretty sure that otherwise the command syntax is correct.

I see that the "xstr" command is passed directly to the HMI with the icon coding. Since the icons are custom built into the HASP HMI SW, is there a way to explicitly list the coding it uses to look up the correct icon? If I knew the exact byte sequence it is looking for, I might have a way to reliably get this into the xstr string.

**Comment by aderusha** [comment]
Can you post some examples that aren't working?  I'm not doing anything at all special with the blueprints, I highlight the icon I want, paste into VS code, and the automation sends the desired character via MQTT.

UTF-8 is the standard encoding for MQTT and is what the panel will expect.  If you look at the page source, [you'll find the codepoint](https://github.com/HASwitchPlate/HASPone/blob/main/images/hasp-fontawesome5.html#L132) used for each of the icons.

**Reply by rebel7580** [reply]
I send this:
```
mqttx: Published hasp/hasp01/command/json, ["p[3].b[6].txt=\"1:08 \"","delay=1","xstr 0,136,50,65,8,8461,0,1,1,3,\"\""]
```
What shows up on the display is 1.08, centered, but no icon.  The "0,136" should put the icon in button 6 ((6-4)*67 +2).
Some icons, like the first one (wineglass) cause an error:
```
{"event":"jsonError","event_source":"nextionParseJson()","event_description":"Failed to parse incoming JSON command with error:IncompleteInput memoryUsage: 78 capacity: 3072"}
```
Most just accept the input, but don't display anything. The pine tree displays an "A"!

So I suspect that just a copy/paste gets mangled by my editor/MQTT publish somehow, which is why I wanted to see if I could get around that by actually typing the codes directly. Just tried that, in various forms, but it just tries to display the first 1 or 2 characters of the code as text.

So I'm about to give up on what I thought would be a simple exercise. :(


BTW, slightly related, did you see  issue #95 ? I think I found a bug in that blueprint while exploring customizing it for different icons. The bug is there on an unmodified blueprint, without any of my changes. Of course, based on my low experience level on your stuff and Home Asisstant in general, I just may be doing something wrong.
